### ISSUE TITLE
Support private Hex packages

### ISSUE THREAD

**Issue by SnakeDoc** [gleam-lang/gleam#3655]
It would be great if Gleam supported publishing and fetching private packages from Hex.pm - https://hex.pm/docs/private

**Comment by lpil** [comment]
We want this for sure.

Could you expand on your need for them? It'll help me prioritise this work.

**Comment by SnakeDoc** [comment]
Thanks for considering this!

The main need for private packages is to allow individuals and organizations to work with internal/proprietary code and libraries that they want to keep private. This feature would enable us to host libraries and code on Hex.pm without making them part of the public registry. This, of course, would enhance Gleam's appeal to business and enterprise.

Presently, it appears the only way to handle private packages is to use local packages, or self-host hex. Neither are very convenient for a variety or reasons.

I have not dug too deeply yet, but it appears we might be able to leverage the work @NthTensor previously did with local packages (and the beginning of git packages). We might be able to add `organization` to `gleam.toml` (and `manifest.toml` where appropriate):

Use the organization when publishing to hex:
```bash
# head of gleam.toml
name = "foo"
organization = "bar"
```

Use the organization when fetching from hex:
```bash
# deps from gleam.toml
[dependencies]
fooey = { version = ">= 1.0.0 and < 2.0.0", organization = "bar" }
```

Users would still be required to also set the appropriate environment variables for `HEXPM_USER` and `HEXPM_PASS`.

Thoughts? 🤔

**Comment by SnakeDoc** [comment]
Yes, we're an early stage startup, so we're just one small data point, but yes, it is currently a bit of a blocker for us. We're exploring work arounds, such as leveraging local packages, but supporting private packages on Hex would be the most ideal option.

**Comment by lpil** [comment]
Great, thank you. Let's try and get this done soon @gleam-lang/compiler-team.

I'm going to be short on time until after CodeBEAM but hopefully we can get it in for the next release.

**Comment by giacomocavalieri** [comment]
Hello! Same for me: I'm at Goto this week and at Code BEAM the next but I can try and tackle this in between the two, or after Berlin! :)

**Comment by lpil** [comment]
I've reached out to Eric to see if we can get a free account to implement this with. If not we'll have to buy some accounts for a month or two.

**Comment by SnakeDoc** [comment]
You guys are awesome 💯

If Hex won't provide free accounts to develop with, please let me know and I'll see what we can do to help.

**Comment by lpil** [comment]
They're sorting it out for us :)

**Comment by dvic** [comment]
@lpil @giacomocavalieri is this currently being worked on? I haven't looked into in the details but we'd really like to have support for this so I'm considering picking this up. But before I proceed I thought I'd check in with you :)

**Comment by lpil** [comment]
Hello! I've done some initial preparatory work but have not been able to work on the private repo support itself. Feel free to work on this one!
